src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentRenderModeDirectivePass.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+#nullable enable
+
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Components;
+
+internal class ComponentRenderModeDirectivePass : IntermediateNodePassBase, IRazorDirectiveClassifierPass
+{
+    protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+    {
+        var @namespace = documentNode.FindPrimaryNamespace();
+        var @class = documentNode.FindPrimaryClass();
+        if (@namespace == null || @class == null)
+        {
+            return;
+        }
+
+        var directives = documentNode.FindDirectiveReferences(ComponentRenderModeDirective.Directive);
+        if (directives.Count == 0)
+        {
+            return;
+        }
+
+        // We don't need to worry about duplicate attributes as we have already replaced any multiples with MalformedDirective
+        Debug.Assert(directives.Count == 1);
+
+        var token = ((DirectiveIntermediateNode)directives[0].Node).Tokens.FirstOrDefault();
+        if (token == null)
+        {
+            return;
+        }
+
+        // generate the inner attribute class
+        // PROTOTYPE: fully qualify type names and extract them out to consts
+        var classDecl = new ClassDeclarationIntermediateNode()
+        {
+            ClassName = "PrivateComponentRenderModeAttribute",
+            BaseType = "RenderModeAttribute",
```

Comment:
Need to use the `global::` qualified name here to reduce chance of ambiguity with locally defined types

Reply:
Yeah I was also covering that in the PROTOTYPE above, but I can make the change now.